Spawn vehicles via RPC

rajat2004 opened the pull request:
Closes #2922

Rebased version of https://github.com/microsoft/AirSim/pull/1937
![airsim-spawn-rpc-2020-01-19](https://user-images.githubusercontent.com/37938604/72684973-4b2ada00-3b0b-11ea-949f-65373a2fc0c5.gif)


https://user-images.githubusercontent.com/37938604/107155117-a93b9100-699c-11eb-8d3b-6c0f27c0c962.mp4

Future improvement possible - pass JSON object as string for the vehicle settings

The example script is very minimal one right now, will need to improve it

Fun fact: I did test with 50 drones, AirSim was able to create them :) but my laptop was definitely struggling to keep up, also had a crash when closing the editor due to RPC error.

Commit:
Support for adding vehicles dynamically at run-time via RPC

Commit:
Support for adding vehicles dynamically at run-time via RPC

KerryMoffittRtn committed:
Fix encoding

KerryMoffittRtn committed:
Fix encoding

Commit:
Sample code, Linux-side

Commit:
Sample code, Linux-side

rajat2004 committed:
Add simAddVehicle() to Python API

rajat2004 committed:
Add simAddVehicle() to Python API

rajat2004 committed:
Add example Python script for spawn vehicle through API

rajat2004 committed:
Add example Python script for spawn vehicle through API

rajat2004 commented:
Need to rebase and test again. One thing I'm not sure about is how the [Airsim.sln](https://github.com/microsoft/AirSim/blob/master/AirSim.sln) file is updated, it seems to indicate it's for Version 15, but we're now using 16 (2019), so doesn't it need to be updated? Or that it doesn't affect the build, etc?

rajat2004 commented:
Working on Linux, Pending Windows .vcxproj update, testing

zimmy87 commented on AirLib/src/api/RpcLibServerBase.cpp:
```diff
@@ -141,8 +142,27 @@ RpcLibServerBase::RpcLibServerBase(ApiProvider* api_provider, const std::string&
 
-    pimpl_->server.
-        bind("simSetVehiclePose", [&](const RpcLibAdapatorsBase::Pose &pose, bool ignore_collision, const std::string& vehicle_name) -> void {
+    pimpl_->server.bind("simAddVehicle", [&](const std::string& vehicle_name, const std::string& vehicle_type, const std::string& pawn_path, float north, float east, float down) -> bool {
```
Could this be simplified by changing the parameters to take in a Pose object instead of 3 coordinates? That way you can get both positional and rotational information.

rajat2004 replied:
Done, thanks for the detailed review!

zimmy87 commented on HelloSpawnedDrones/HelloSpawnedDrones.cpp:
```diff
@@ -0,0 +1,172 @@
+
+	try
+	{
+		msr::airlib::MultirotorRpcLibClient *client = new msr::airlib::MultirotorRpcLibClient(host, port, timeout_s);
```
You should be able to reference MultirotorRpcLibClient without specifying the msr::airlib namespace since you have a using statement above.

rajat2004 replied:
Fixed

zimmy87 commented on HelloSpawnedDrones/HelloSpawnedDrones.cpp:
```diff
@@ -0,0 +1,172 @@
+
+		z -= ordinal * altitudeDelta;
+		float timeout = 10.0f;
+		client->moveToZAsync(z, speed, timeout, msr::airlib::YawMode(), -1.0f, 1.0f, vehicleName)->waitOnLastTask();
```
Same as with MultirotorRpcLibClient, you should be able to reference YawMode without specifying the msr::airlib namespace.

rajat2004 replied:
Fixed

zimmy87 commented on HelloSpawnedDrones/HelloSpawnedDrones.cpp:
```diff
@@ -0,0 +1,172 @@
+		std::cout << "Confirming connections..." << std::endl;
+		client->confirmConnection();
+
+		std::string vehicleName = "UAV_" + std::to_string(ordinal);
```
From looking at HelloDrone and DroneShell, it looks like these standalone projects are still conforming to the same naming conventions as the rest of the repo, so this variable should be under_scored and not camelCase. This applies to a lot of local variables in this file, so I'll use this comment to track them all.

rajat2004 replied:
Changed to `vehicle_name`, will slowly work on the others as well. Some of them are in camelCase in the HelloDrone as well, so might keep those as is

zimmy87 reviewed (changes requested):
Thanks for creating this PR @rajat2004! Once you're ready to submit this PR, I can take another look at this PR in case you've already addressed some of these comments in your next update.

rajat2004 committed:
Use Pose instead of NED values, fix missing override

rajat2004 committed:
Use Pose instead of NED values, fix missing override

rajat2004 committed:
Fix HelloSpawnedDrones C++ file to use new API, naming convention, convert to spaces

rajat2004 committed:
Fix HelloSpawnedDrones C++ file to use new API, naming convention, convert to spaces

rajat2004 commented on HelloSpawnedDrones/HelloSpawnedDrones.cpp:
```diff
@@ -0,0 +1,172 @@
+
+	try
+	{
+		std::cout << "Press Enter to begin..." << std::endl; std::cin.get();
```
Fixed

rajat2004 commented on AirLib/include/api/RpcLibClientBase.hpp:
```diff
@@ -94,6 +94,7 @@ class RpcLibClientBase {
     vector<uint8_t> simGetImage(const std::string& camera_name, ImageCaptureBase::ImageType type, const std::string& vehicle_name = "");
 
     vector<MeshPositionVertexBuffersResponse> simGetMeshPositionVertexBuffers();
+    bool simAddVehicle(const std::string& vehicle_name, const std::string& vehicle_type, const std::string& pawn_path, float north, float east, float down);
```
I actually didn't do much work on this apart for rebasing and adding the Python APIs. So mostly everything is the same as was in the original PR. Updated the accept pose instead of NED values.

rajat2004 committed:
Consolidate vehicle creation code in SimModeBase

rajat2004 committed:
Consolidate vehicle creation code in SimModeBase

rajat2004 commented on AirLib/include/common/AirSimSettings.hpp:
```diff
@@ -414,6 +414,15 @@ struct AirSimSettings {
+        std::unique_ptr<VehicleSetting> vehicle_setting_p = std::unique_ptr<VehicleSetting>(new VehicleSetting());
+
+        // Usually we have a pointer to an entry from the json, but here we have to make a new one
+        vehicles[vehicle_setting.vehicle_name] = std::move(vehicle_setting_p);
```
This is turning out to be a bigger rabbit hole than I thought. The map stores `unique_ptr`, and the object passed by ref can be used to create a unique ptr, but that's a [bad thing to do](https://stackoverflow.com/questions/28502067/how-to-assign-the-address-of-an-existing-object-to-a-smart-pointer). The object is created in [WorldSimApi.cpp](https://github.com/microsoft/AirSim/pull/2390/files#diff-89d4ec9b62486b1322e5ba2dd9936b13962f9ed113ec5e35a0678846889c7e2dR274) on the stack, so the object would be erased after the function returns and the ptr will refer to non-existing object, therefore using `reset` seems to be out.

Creating a `new VehicleSetting(vehicle_setting)` i.e using a copy-constructor or even move-constructor also doesn't work since the struct contains a map of `unique_ptr` for SensorSettings, which doesn't allow this. The simplest way I can think of right now is to add a method in `AirSimSettings.hpp` to create the object on heap like being done when VehicleSettings are loaded, and then in SimModeBase, a ref to that object can be used, which should avoid these problems. But will need to think on this a bit more

rajat2004 replied:
Fixed in https://github.com/microsoft/AirSim/commit/ae7fb5df8083472820505074841a00236a78d1aa

rajat2004 commented:
@zimmy87 and others also, If possible, could you test this on Windows as well, I'll also try to do so on my end. The changes in AirSim.sln & .vcxproj files are something which I have no idea about, these were already present in the old PR and have been carried forward here.

rajat2004 committed:
Store correct vehicle settings instead of default one

rajat2004 committed:
Store correct vehicle settings instead of default one

rajat2004 committed:
Unity: Add unimplemented simAddVehicle API method

rajat2004 committed:
Unity: Add unimplemented simAddVehicle API method

rajat2004 commented:
I'm marking this PR as Ready for review since the first review comments as well as Unity non-implemented method has been tested, plus the HelloSpawnedDrones and the Python script (See video in PR description for the behaviour of HelloSpawnedDrones). Additionally, I've tested this PR on top of #2936 which adds an APi to get the list of current vehicle names. Somewhat useful in normal cases, and more so when combined with this PR. After running `./HelloSpawnedDrones 5`, the API -

```
>>> cl.listVehicles()
['SimpleFlight', 'UAV_0', 'UAV_1', 'UAV_2', 'UAV_3', 'UAV_4']
```

rajat2004 committed:
HelloSpawnedDrones sample (Windows), and note in docs

rajat2004 committed:
HelloSpawnedDrones sample (Windows), and note in docs

rajat2004 commented:
@zimmy87 If possible, could you have a look at this PR? Specifically, I'm having some trouble in fixing the HelloSpawnedDrones Visual Studio project, it needs to be updated, and the project properties such as include folders, libraries, etc are not set I think. It should also be added to AirSim.sln file

rajat2004 committed:
Adapt to AirSim naming conventions, cleanup

rajat2004 committed:
Adapt to AirSim naming conventions, cleanup

zimmy87 committed:
Updating AirSim solution for new project HelloSpawnedDrones

This change adds the HelloSpawnedDrones.vcxproj to AirSim.sln and also sets output and intermediate directories for the new project

zimmy87 commented on docs/apis_cpp.md:
```diff
@@ -88,5 +88,6 @@ int main()
 ## See Also
 * [Examples](../Examples) of how to use internal infrastructure in AirSim in your other projects
 * [DroneShell](../DroneShell) app shows how to make simple interface using C++ APIs to control drones
+* [HelloSpawnedDrones](./HelloSpawnedDrones) app shows how to make additional vehicles on the fly
```
shouldn't the hyperlink for HelloSpawnedDrones be `../HelloSpawnedDrones`?

rajat2004 replied:
Nice catch, fixed!

zimmy87 reviewed (changes requested):
I tested this PR locally and it works for me, so I think we'll be close to merging once the remaining comments are addressed and we resolve the updates to project files

zimmy87 commented:
Hi @rajat2004, as per updating HelloSpawnedDrones.vcxproj and HelloSpawnedDrones.vcxproj, I've made a update to both these files in a commit [here](https://github.com/zimmy87/AirSim/commit/0fa412efadbefdd35b8b6671a12bcc3f16989fe0) in my own AirSim fork that you can feel free to copy into this PR. With this commit, I'm able to build HelloSpawnedDrones from the AirSim solution locally, however this is breaking the Windows build check for me, so I'm still missing some dependencies in this commit. I will try and spend more time fixing the Windows build, and keep you updated once I've reached some resolution.

rajat2004 committed:
HelloDrone: Fix variable naming to align with convention

rajat2004 commented on HelloSpawnedDrones/HelloSpawnedDrones.cpp:
```diff
@@ -0,0 +1,174 @@
+        auto groundPosition = client->getMultirotorState(vehicle_name).getPosition();
+        float groundZ = groundPosition.z(); // current position (NED coordinate system).  
+
+        float takeoffTimeout = 5;
```
Done, also added https://github.com/microsoft/AirSim/pull/2390/commits/ede14593f51f904c4206d201cbb326107d01e609 which makes the changes in HelloDrone

rajat2004 commented on HelloSpawnedDrones/HelloSpawnedDrones.cpp:
```diff
@@ -0,0 +1,174 @@
+        float z = position.z(); // current position (NED coordinate system).  
+
+        // Altitude difference between each platform, in meters
+        const float altitudeDelta = 1.0f;
```
Fixed

rajat2004 commented on HelloSpawnedDrones/HelloSpawnedDrones.cpp:
```diff
@@ -0,0 +1,174 @@
+
+        const float size = 5.0f;
+        duration = size / speed;
+        DrivetrainType driveTrain = DrivetrainType::ForwardOnly;
```
Fixed

rajat2004 commented on HelloSpawnedDrones/HelloSpawnedDrones.cpp:
```diff
@@ -0,0 +1,174 @@
+        std::vector<std::thread> clientThreads;
+
+        // Count down, so the first one can easily go the highest (without knowing count)
+        int clientOrdinal = num_platforms - 1;
```
Fixed

rajat2004 commented on Unity/AirLibWrapper/AirsimWrapper/Source/WorldSimApi.cpp:
```diff
@@ -195,4 +195,12 @@ void WorldSimApi::setWind(const Vector3r& wind) const
     simmode_->setWind(wind);
 };
 
+bool addVehicle(const std::string& vehicle_name, const std::string& vehicle_type, const WorldSimApi::Pose& pose, const std::string& pawn_path)
```
Fixed, the problem was the missing `WorldSimApi::`

rajat2004 commented on HelloSpawnedDrones/HelloSpawnedDrones.cpp:
```diff
@@ -0,0 +1,174 @@
+    if (argc > 1)
+    {
+        std::cout << "Num plats string: " << argv[1] << std::endl;
+        int numPlatsInt = atoi(argv[1]);
```
Removed the var itself, so fixed :)

rajat2004 committed:
Docs: Fix HelloSpawnedDrones location

rajat2004 commented:
@zimmy87 Sure, will bring that commit in later. I think it might be best to do that after all the problems have been figured out so can be done in one swoop. Once it's fixed, let me know and I can cherry-pick the commits onto my branch, or you could push to mine as well since maintainers can do that

zimmy87 committed:
Update HelloSpawnedDrones.vcxproj

add project dependencies

zimmy87 committed:
Merge branch 'rajat2004-pr/1937' into pr/1937

zimmy87 committed:
Update SimModeBase.h

remove duplicate declaration of registerPhysicsBody()

zimmy87 commented:
Tested latest version locally on Windows & Ubuntu and it's working & building just fine for me, so I am moving ahead with merging.

rajat2004 commented:
@zimmy87 Thanks for the Windows fixes and merging! I think the Python script can be improved further, and probably should add a bit of info in the docs as well, will do so in another PR

rajat2004 commented:
@Roopesh-Bharatwaj-K-R Are you using the latest master rather than some other branch that isn't up to date and doesn't include this PR? Also, did you use `import setup_path` before doing `import airsim`? This is required to use the airsim module from local repo, rather than the pip installed one.

Roopesh-Bharatwaj-K-R commented:
@rajat2004 I Used import setup_path before import airsim. I'm not sure about the master.please give more details on this.
